src/tree.rs

```diff
@@ -74,61 +235,236 @@ impl Tree {
     }
 
     pub fn guids<'t>(&'t self) -> impl Iterator<Item = &Guid> + 't {
-        self.index_by_guid
-            .keys()
+        assert_eq!(self.entries.len(), self.entry_index_by_guid.len());
+        self.entries.iter()
+            .map(|entry| &entry.item.guid)
             .chain(self.deleted_guids.iter())
     }
 
+    /// Returns the node for a given `guid`, or `None` if a node with the `guid`
+    /// doesn't exist in the tree, or was deleted.
     pub fn node_for_guid(&self, guid: &Guid) -> Option<Node> {
-        self.index_by_guid
-            .get(guid)
-            .map(|index| self.node(*index))
+        assert_eq!(self.entries.len(), self.entry_index_by_guid.len());
+        self.entry_index_by_guid.get(guid).map(|&index| {
+            let entry = &self.entries[index];
+            if self.orphan_indices_by_parent_guid.contains_key(&entry.item.guid) {
+                // Check if this node has reparented orphans (rule 2).
+                return Node(self, entry, Divergence::Diverged);
+            }
+            if entry.is(&self.entries[self.reparent_orphans_to_default_index()]) {
+                // If this node is the default folder for reparented orphans,
+                // check if we have any remaining orphans that reference
+                // nonexistent or non-folder parents (rule 3).
+                let needs_reparenting = |guid| {
+                    match self.entry_index_by_guid.get(guid) {
```

Comment:
nit: map_or's maybe an improvement
```suggestion
                    self.entry_index_by_guid
                        .get(guid)
                        .map_or(true, |&index| !self.entries[index].item.is_folder())
```